### PR TITLE
feat(voice): voice dictation across chat, wardrobe search, and add-item form (#152)

### LINKED ISSUE #152: Voice dictation — add clothing by speaking + voice search
## Problem

Adding a wardrobe item today is a 10+ tap form: brand, item name, color, garment type chip, category, season chip, style chip, fabric, size, optional price, image picker. Even with bulk-upload + auto-tag, manual edits and corrections are still finger-heavy.

Voice dictation would shrink this to a single utterance — \"It's a navy wool overcoat from Lemaire, fall/winter, business casual\" — and let the user search the wardrobe by speaking instead of typing.

### A. Voice add — manual-entry form
On `/wardrobe/new` (and the manual-fill route from #112), a mic button next to the form. User dictates once; we transcribe via Whisper or expo-speech-recognition, then route through Gemini structured-output (same VisionTagResult schema as vision_tagger) to extract the fields.

The fields it fills:
- brand, item_name, color (free text — direct from transcript)
- garment_type, garment_category, season, style (enum — Gemini picks the matching enum)
- fabric, size (free text — only if mentioned)

Vision-only mode and voice-only mode both feed the same eventual row. Voice + photo together (\"this is the navy coat I just shot\") should be the magic path.

### B. Voice search — wardrobe grid + chat
On `/wardrobe`, mic button on the search bar. \"Show me everything olive and casual\" → set the search filter. Same on the chat tab for outfit queries — \"What should I wear to dinner tomorrow?\" should accept voice input.

## Open questions

- Provider: native expo-speech-recognition (free, on-device on iOS, mediocre on Android web) vs Whisper API via the agent (more accurate, costs cents per minute, network-dependent).
- Permissions UX: first-time mic-permission prompt should be on tap-to-speak, not on app launch.
- Transcript review: do we show the transcript before posting, or auto-submit on silence detection? Probably show + edit on add, auto-submit on search.

## Out of scope (separate issues)

- Voice replies (TTS) — this issue covers input only.
- Voice on the iOS keyboard — that's already a system-level input method; we just need to not break it on `<Input />` components.

Parent: epic to be assigned.

### PR DESCRIPTION
**Pull request by William-Hill**
Closes #152.

Yes — voice works on the **chat** surface, plus the two surfaces called out in the issue. Three tap-to-talk entry points share one hook + one mic primitive:

## Surfaces
- **Chat input** (B2): mic next to `SEND`. Transcript appends to whatever's typed; explicit SEND keeps the user in control of agent calls.
- **Wardrobe screen** (B1): new editorial search bar with a mic pill. Transcript replaces the query and triggers the existing `getWardrobeItems({ search })` filter.
- **Item form** (A): voice-fill banner near the top of `/wardrobe/new`. Transcript posts to a new `/wardrobe/parse-voice` endpoint that runs Gemini structured output and returns nullable wardrobe fields. Only non-null fields overwrite form state, so partial dictation merges with any typing the user has already done.

## Architecture
- `useVoiceDictation` hook wraps `expo-speech-recognition` with a single tap-to-talk surface. Permission is requested on first **start**, not on mount.
- `MicButton` UI primitive (compact + pill variants) with a pulse animation while listening.
- `server/lib/voice_parser.py` — Gemini 2.5 Flash with `VoiceParseResult` schema that mirrors the `vision_tagger.py` enums. Free-text fields stay nullable so the model omits anything the user didn't say instead of hallucinating it. 1500-char cap on inbound transcripts.
- `POST /wardrobe/parse-voice` route on the existing wardrobe router; auth-gated via `current_user_id`.
- `tests/test_voice_parser.py` — validation, schema parsing, fallback-from-text path, unparseable-response error mapping. 5 tests, all passing.

## Provider choice
Per the issue's open question: device-only via `expo-speech-recognition`. iOS uses on-device recognition; Android/web fall back to the platform default. Whisper API fallback is deferred — easy to add later as a runtime switch in the hook.

## Permissions
- `microphonePermission` and `speechRecognitionPermission` strings registered as a config-plugin block in `app.json`. Requires a dev-client rebuild — Expo Go won't work for this feature.

## Out of scope (per the issue)
- Voice replies (TTS).
- Whisper API fallback.

## Test plan
- [ ] Build a dev client (`npx expo prebuild && eas build --profile development`) and verify mic permission prompt fires on first tap, not on app launch.
- [ ] Chat: tap mic, dictate "What should I wear to dinner tomorrow?", verify transcript lands in the input and SEND fires the agent.
- [ ] Wardrobe: search "show me everything olive and casual", verify the grid filters.
- [ ] Item form: dictate "Navy wool overcoat from Lemaire, fall winter, business casual" and verify all six fields populate. Confirm low-confidence transcripts show the soft toast.
- [ ] Pytest: `pytest tests/test_voice_parser.py` (passes locally — 5/5).

## Summary by CodeRabbit

* **New Features**
  * Voice dictation across the app: chat input, wardrobe search, and a “voice fill” for creating items (dictation appends to typed text).
  * Text search for wardrobe with debounced queries.
  * Microphone button UI with animated listening state and accessible labels.
  * Server endpoint to parse voice transcripts into structured wardrobe fields.

* **Chores**
  * Added platform speech-recognition configuration and dependency.

* **Tests**
  * New unit tests covering voice parsing and edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/William-Hill/ptso_agent/pull/164)

### CODE REVIEW

File: server/routers/wardrobe.py [hunk 1]
```diff
@@ -24,6 +25,37 @@ class AutoTagResponse(BaseModel):
...
+):
+    """Parse a spoken garment description into wardrobe-item fields.
+
+    Returns nullable fields shaped like VisionTagResult so the manual-entry
+    form can pre-fill itself with whatever the user mentioned. Fields the
+    user didn't say come back null — the form leaves those at their existing
+    defaults instead of guessing.
+    """
+    try:
+        fields, confidence = parse_transcript(req.transcript)
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply server/routers/wardrobe.py]: Fixed in 18a3037 — `parse_voice` now runs `parse_transcript` via `asyncio.to_thread` so the blocking Gemini call no longer stalls the event loop.

File: mobile/src/components/voice/mic-button.tsx [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+          withTiming(1.25, { duration: 480 }),
+          withTiming(1, { duration: 480 })
+        ),
+        -1,
+        true
+      );
+    } else {
+      pulse.value = withTiming(1, { duration: 200 });
+    }
+  }, [isListening, pulse]);
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/src/components/voice/mic-button.tsx]: Fixed in 18a3037 — dropped `pulse` from the useEffect dependency array.

File: mobile/src/components/voice/mic-button.tsx [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            backgroundColor: tone,
+          },
+          dotStyle,
+        ]}
+      />
+      <View>
+        <Text variant="overline" style={{ color: tone, letterSpacing: 2 }}>
+          {label}
+        </Text>
+      </View>
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/src/components/voice/mic-button.tsx]: Fixed in 18a3037 — removed the redundant `View` wrapper; rendering `Text` directly.

File: mobile/src/lib/voice/use-voice-dictation.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+   */
+  onError?: (code: string, message?: string) => void;
+}
+
+/**
+ * Wraps expo-speech-recognition behind a single tap-to-talk hook. Permission
+ * is requested on first start (not on hook mount) so users only see the prompt
+ * when they actually try to dictate. Handles platform unsupported gracefully
+ * (returns status='unsupported' instead of throwing) so callers can hide UI.
+ */
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/src/lib/voice/use-voice-dictation.ts:39]: Fixed in 18a3037 — `start()` now short-circuits to `status='unsupported'` when `ExpoSpeechRecognitionModule.isRecognitionAvailable()` is false, and both the catch block and the `error` event listener map `service-not-allowed` / `language-not-supported` codes to `'unsupported'` instead of `'error'`.

File: mobile/src/lib/wardrobe/parse-voice-client.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      throw new Error(
+        `parse-voice failed (${response.status}): ${body.slice(0, 200)}`
+      );
+    }
+
+    return (await response.json()) as ParseVoiceResult;
+  } finally {
+    clearTimeout(timeoutId);
+  }
+}
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/src/lib/wardrobe/parse-voice-client.ts:75]: No change needed — the response shape is owned by our own FastAPI server (`ParseVoiceResponse` Pydantic model), and the client and server types are versioned together. Adding a Zod runtime check would duplicate validation that's already enforced server-side without catching anything actionable.

File: mobile/src/lib/voice/use-voice-dictation.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  });
+
+  useSpeechRecognitionEvent('error', (event) => {
+    if (event.error === 'service-not-allowed' || event.error === 'language-not-supported') {
+      setStatus('unsupported');
+    } else {
+      setStatus('error');
+    }
+    errorRef.current?.(event.error ?? 'unknown', event.message);
+  });
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/src/lib/voice/use-voice-dictation.ts:98]: Fixed in 077ef18 — the `error` listener now short-circuits on `event.error === 'aborted'` (the documented response to `cancel()`/`abort()`), resets to `idle`, and skips the error callback.

File: mobile/src/lib/voice/use-voice-dictation.ts [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+      }
+      ExpoSpeechRecognitionModule.start({
+        lang,
+        interimResults: true,
+        continuous: false,
+        // On iOS, on-device recognition is preferred when available — keeps
+        // dictation private and fast. Falls back to network when the locale
+        // isn't installed locally.
+        requiresOnDeviceRecognition: Platform.OS === 'ios',
+        addsPunctuation: true,
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/src/lib/voice/use-voice-dictation.ts]: Fixed in 077ef18 — on iOS we now call `requestMicrophonePermissionsAsync()` (since `requiresOnDeviceRecognition` is set), and fall back to `requestPermissionsAsync()` only on non-iOS platforms.

File: server/routers/wardrobe.py [hunk 8]
```diff
@@ -24,6 +26,37 @@ class AutoTagResponse(BaseModel):
     results: list[AutoTagResult]
 
 
+class ParseVoiceRequest(BaseModel):
+    transcript: str = Field(..., min_length=1, max_length=MAX_TRANSCRIPT_CHARS)
+
+
+class ParseVoiceResponse(BaseModel):
+    fields: dict
+    confidence: float
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply server/routers/wardrobe.py]: Fixed in 077ef18 — split `VoiceParseResult` into a `VoiceParseFields` base (no confidence) and the result type that extends it; `ParseVoiceResponse.fields` is now typed as `VoiceParseFields` and `confidence` is bounded with `Field(..., ge=0.0, le=1.0)`.

File: server/routers/wardrobe.py [hunk 9]
```diff
@@ -24,6 +26,37 @@ class AutoTagResponse(BaseModel):
...
+    user didn't say come back null — the form leaves those at their existing
+    defaults instead of guessing.
+    """
+    try:
+        fields, confidence = await asyncio.to_thread(parse_transcript, req.transcript)
+    except ValueError as e:
+        raise HTTPException(status_code=400, detail=str(e)) from e
+    except RuntimeError as e:
+        logger.exception("Voice parse failed")
+        raise HTTPException(status_code=502, detail=str(e)) from e
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply server/routers/wardrobe.py]: Fixed in 077ef18 — the 502 handler now returns a generic "Voice parsing service is temporarily unavailable." message; full details remain in logs via `logger.exception`.

File: mobile/src/lib/voice/use-voice-dictation.ts [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+    } catch (e: any) {
+      const code = e?.code;
+      if (code === 'service-not-allowed' || code === 'language-not-supported') {
+        setStatus('unsupported');
+      } else {
+        setStatus('error');
+      }
+      errorRef.current?.('start_failed', e?.message ?? String(e));
+    }
+  }, [lang]);
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/src/lib/voice/use-voice-dictation.ts:147]: Fixed in b5a67ed — added a `startInFlightRef` re-entrancy guard. `start()` returns early if a session is already in flight, and the flag is cleared on every termination path (catch, denial, unsupported, `error` event, `end` event).

File: mobile/src/lib/voice/use-voice-dictation.ts [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+
+  const cancel = useCallback(() => {
+    try {
+      ExpoSpeechRecognitionModule.abort();
+    } catch {
+      // No-op.
+    }
+    setStatus('idle');
+    setTranscript('');
+  }, []);
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/src/lib/voice/use-voice-dictation.ts:169]: Fixed in b5a67ed — `cancel()` now invokes `partialRef.current?.('')` so any external transcript preview (e.g. add-item form draft) clears alongside the hook's internal state.

File: server/lib/voice_parser.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+    )
+
+    parsed: VoiceParseResult | None = getattr(response, "parsed", None)
+    if parsed is None:
+        try:
+            parsed = VoiceParseResult.model_validate_json(response.text or "")
+        except ValidationError as e:
+            logger.warning("Voice parser returned unparseable response: %s", e)
+            raise RuntimeError("Gemini response did not match VoiceParseResult schema.") from e
+
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply server/lib/voice_parser.py]: Fixed in b5a67ed — wrapped `client.models.generate_content(...)` in try/except that logs via `logger.exception` and re-raises as `RuntimeError` so SDK/network/auth failures honor the documented contract and the router maps them to 502.

File: server/lib/voice_parser.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+    confidence = parsed.confidence
+    fields = parsed.model_dump(exclude={"confidence"})
+    return fields, confidence
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply server/lib/voice_parser.py]: Fixed in 80b4cbf — wrapped the entire `parse_transcript` body in `try/finally` so `client.close()` runs on success, `generate_content` failures, and parse failures alike.
